### DISCUSSION TITLE
Possible to install `truffleruby+graalvm@latest` to keep it updated?

### DISCUSSION THREAD

**Discussion by ylluminate** [jdx/mise#2498] [Q&A]
Is there a way to install `truffleruby+graalvm@latest` in a manner that would allow mise updates to keep it continually up-to-date?

**Comment by Marukome0743** [comment]
Flavoured Ruby version requests are now supported. This was implemented in #6533 and released in mise v2025.10.3.

You can keep the latest TruffleRuby + GraalVM release selected with:

```sh
mise use ruby@truffleruby+graalvm
```

The resulting configuration is equivalent to:

```toml
[tools]
ruby = "truffleruby+graalvm"
```

Because this remains a flavour request rather than an exact version, `mise upgrade ruby` can resolve and install a newer matching release when one becomes available.

I verified this on the current version of mise. An installation dry run resolved the configuration above to:

```text
ruby@truffleruby+graalvm-34.0.1
```

That was the latest matching version available at the time of testing.

*AI-assisted — Tool: Codex; model: OpenAI/GPT-5; version: unavailable.*

**Reply by rkh** [reply]
The logic seems to be broken. `ruby@truffleruby` now always picks up the `graalvm` flavor for me. How can is specify that I want the latest non-dev native version of truffleruby?

**Reply by Marukome0743** [reply]
You're right, and it's a regression — nothing wrong with your setup. Thanks for reporting it.

### What was happening

mise's fuzzy version matcher accepted `+` as a separator directly after the query:

```rust
Regex::new(&format!("^{query_pattern}([+\\-.].+)?$")).unwrap()
```

So `truffleruby` matched `truffleruby-34.0.1` **and** `truffleruby+graalvm-34.0.1`, and when several versions match, mise takes the last one in the list. ruby-build lists the graalvm builds after the native ones, so graalvm won every time.

Up to v2025.12.9 the pattern was `^{query}([-.].+)?$` and `ruby@truffleruby` resolved to the native build. `+` entered the character class in #7332 (v2025.12.10), whose subject was vendor prefixes like `temurin-` — its description quoted the previous pattern as `^{query}([+-.].+)?$` when the code read `^{query}([-.].+)?$`, so the `+` looks like it was carried over by mistake rather than intended.

It also contradicted a rule mise already applies elsewhere: a bare vendor or flavour name is supposed to select the base flavour, which is why `java@liberica` gives `liberica` and not `liberica-nik`.

### Fix

#12118 is merged into `main`. `+` is now only treated as a separator for numeric queries, where it introduces semver build metadata (`1.9.1` → `1.9.1+hotfix.2`); for name queries it no longer crosses into a different flavour.

```console
$ mise latest ruby@truffleruby
truffleruby-34.0.1

$ mise latest ruby@truffleruby+graalvm
truffleruby+graalvm-34.0.1
```

Java, Flutter's `+hotfix` versions, and every other tool resolve exactly as before — the two backends with the most tools, `aqua:` and `ubi:`, already excluded `+` from their separator class, so this only ever affected the shared matcher.

The e2e test that should have caught this asserted only that `mise latest ruby@truffleruby` ends in `-N.N.N`, which `truffleruby+graalvm-34.0.1` also satisfies. It now pins the flavour, so the same regression can't slip through again.

This landed after v2026.8.8, so it will be in the next release.

### Until then

Adding a trailing hyphen restricts the query to the native series:

```sh
mise use ruby@truffleruby-
```

That resolves to `truffleruby-34.0.1` today. It also answers the non-dev part of your question: `truffleruby-dev` and the old `1.0.0-rc*` builds are treated as pre-releases and are not selected.

*AI-assisted — Tool: Claude Code; model: Anthropic/claude-opus-5; version: 2.1.222.*
